# Update Null Values

In [1]:
import pandas as pd
import numpy as np

## Load Data

In [2]:
data = pd.read_csv('../final_project.csv')

## Locate Null Values: Column by Row - Populate Dict

Column by row where value is null

In [3]:
data_notnull = data.notnull()
nullValues = {}
for column in data:
    nullValues.update({column: data_notnull[data_notnull[column]==False].index.to_list()})

Display the ten index's (Rows) where there is missing values (NaN) in attribute x0.

In [8]:
nullValues['x0'][0:10]

[1282, 2290, 7491, 15484, 22497, 23192, 28954, 35999, 53729, 72318]

Proving there are NaN's within the first ten locations within the x0 column

In [9]:
# data.loc[ROW, COLUMN]
data.loc[nullValues['x0'][0:10] ,'x0']

1282    NaN
2290    NaN
7491    NaN
15484   NaN
22497   NaN
23192   NaN
28954   NaN
35999   NaN
53729   NaN
72318   NaN
Name: x0, dtype: float64

## Distinct Rows With Missing Value

In [10]:
unique_index = []
for key, value in nullValues.items():
    for i in value:
        if i not in unique_index: 
            unique_index.append(i)
unique_index.sort()

In [11]:
len(unique_index)

1608

## Replace NaN With arithmetic averages (Mean, Median, Mode, etc)

In [14]:
for column in data.columns:
    print(column, data[column].mode()[0])

x0 -1.5926345910622286
x1 -26.278302059686013
x2 -59.394048122248535
x3 -35.47659439723979
x4 -28.46753554577105
x5 -33.82298833073262
x6 -86.35448303441738
x7 -181.50697640168946
x8 -37.69104519154447
x9 -27.98065918344565
x10 -36.30657084793107
x11 -38.09286935060288
x12 -64.19796679643913
x13 -38.7235141494444
x14 -30.905214449583514
x15 -17.00235932515357
x16 -26.042982783860353
x17 -34.395897860607484
x18 -20.198685507387676
x19 -35.63339624523899
x20 -26.677395943693963
x21 -43.501854323465565
x22 -23.644193432638946
x23 -66.64034062402885
x24 asia
x25 -6.364653190718416
x26 -3.857483547465956
x27 -32.00355530339034
x28 -72.8967054221028
x29 July
x30 wednesday
x31 -12.289364349178767
x32 0.01%
x33 -7.451454454619068
x34 -36.11660589927189
x35 -10.008149204275279
x36 -6.866023757082137
x37 $-311.26
x38 -74.2975589171604
x39 -22.101647306158306
x40 -74.0591964973672
x41 -82.16722406683455
x42 -27.93374955024711
x43 -6.876233824751421
x44 -17.983487481796264
x45 -1.7532213009894346


What is the mode for column x0

In [10]:
data_copy['x0'].mode()

0   -1.592635
dtype: float64

Return to nan within the first location (row 1282) within the x0 column to see that it has been replace with the mode

In [12]:
data_copy.loc[nullValues['x0'][0] ,'x0']

-1.5926345910622286